### ISSUE TITLE
[stable/mongodb-replicaset] How do I enable Authentication?

### ISSUE THREAD

**Issue by shinebayar-g** [helm/charts#8657]
**Is this a request for help?**: Help

---

**Which chart**: [stable/mongodb-replicaset]

How do I enable MongoDB authentication?
I'm using this command to deploy MongoDB.
```
helm install --name mongodb-replicaset -f values.yaml stable/mongodb-replicaset
```
Here is my values.yaml file
```YAML
replicas: 3
port: 27017

replicaSetName: rs0

podDisruptionBudget: {}
  # maxUnavailable: 1
  # minAvailable: 2

auth:
  enabled: true
  adminUser: admin
  adminPassword: passwordDB1231231123zzdd19
  # metricsUser: metrics
  # metricsPassword: password
  key: 'mpJJByZbcEsv/9YoTSSZSSo6OXMHw8XYWFS+O5Jp1sz2iaCfbFvnsGeQcsU/vBGN
edQS4IbBFyniYD/Jn7Gu4g8qyXTYwYaALgCfvnnJlhz0z1cDjV3HRPBmED4EBIEv
t+dHdsZeMbBt/gZsZIhuWC+IPfZaq3k5azuysgwlC0AiJFVPCSB4E8XxcVvfv+hQ
Lg0vE/asvvKf5hg3PZCR+Rot9HE6BAInj24uRFXAVRnphs411qRNeOw8enrLvUVi
PSm+t+1J6zhLQ+LlZxG7fbFd8Ke+8nSGIt+raiXRhCLS9/MIqhcjF6UWObzQB/6o
zfQOFRgTV0cy5FJnyEgVq0rIg/NZYKArozxei1Th+Cv6jaMGBl+Jvn/9HyRMsSQX
VQsmQYvvLW56WUyFfhGB61deQRh+1JZoWOLzlB2422nvkW70C1UDC4lnTEK2WxIK
kno8tp3sIh+WVwCnOIGnZQKE0+AYuF0VVWPMFCPirdCw87TfSRtNW7ka/ivTT6LS
svhV4k1MisDOUS2miO87yqLWf8vQvwfV0J6S0wuNPVR4FPhYMjNl0FK+pNTScZMO
AuEQV9zQTKIiUgyop7Ch95zNNwsgq1tTLIMbUUzMK6J+dpbooU5zD+rME5k50V7O
E+IKEs0YodFnVMQGvnNCB0L3Wtz0qKApsIJUpqNRcJXAOui9XGlXe9fQxAeZqRJ4
TybJNiwUPcGnD4EQZI0VajfVqAnpZPQ9RKtZi0RmVo9xzD0YbALsIY4H92jrNmDW
9v+1XI1oIUJtWfa9kdZCsgfjWFeV27os4WouVL3mBPpBGb3ESs5kDVAV6WYkmjcj
qTPLQDXJc5abf3d2HVWXWN7qCP8TqEbg+c3BU9hZXdgCBWaelfCtNmicIygSgzsL
gLbN3x8fdv6Bzx16txAnp9OsEn+mk5cm7KsDPCr9f85pW7Eq/bdvqAjZyrhT/ILw
g6t8AOI2k6wie3Y6EaYRDEnJKud5uyp/U+szFykaro6wKx5y'
  # existingKeySecret:
  # existingAdminSecret:
  # exisitingMetricsSecret:

# Specs for the Docker image for the init container that establishes the replica set
installImage:
  repository: k8s.gcr.io/mongodb-install
  tag: 0.6
  pullPolicy: IfNotPresent

# Specs for the MongoDB image
image:
  repository: mongo
  tag: 3.6
  pullPolicy: IfNotPresent

# Additional environment variables to be set in the container
extraVars: {}
# - name: TCMALLOC_AGGRESSIVE_DECOMMIT
#   value: "true"

# Prometheus Metrics Exporter
metrics:
  enabled: false
  image:
    repository: ssalaues/mongodb-exporter
    tag: 0.6.1
    pullPolicy: IfNotPresent
  port: 9216
  path: "/metrics"
  socketTimeout: 3s
  syncTimeout: 1m
  prometheusServiceDiscovery: true
  resources: {}

# Annotations to be added to MongoDB pods
podAnnotations: {}

securityContext:
  runAsUser: 999
  fsGroup: 999
  runAsNonRoot: true

resources: {}
# limits:
#   cpu: 500m
#   memory: 512Mi
# requests:
#   cpu: 100m
#   memory: 256Mi

## Node selector
## ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector
nodeSelector: {}

affinity: {}

tolerations: []

extraLabels: {}

persistentVolume:
  enabled: true
  ## mongodb-replicaset data Persistent Volume Storage Class
  ## If defined, storageClassName: <storageClass>
  ## If set to "-", storageClassName: "", which disables dynamic provisioning
  ## If undefined (the default) or set to null, no storageClassName spec is
  ##   set, choosing the default provisioner.  (gp2 on AWS, standard on
  ##   GKE, AWS & OpenStack)
  ##
  # storageClass: "-"
  accessModes:
    - ReadWriteOnce
  size: 10Gi
  annotations: {}

# Annotations to be added to the service
serviceAnnotations: {}

tls:
  # Enable or disable MongoDB TLS support
  enabled: false
  # Please generate your own TLS CA by generating it via:
  # $ openssl genrsa -out ca.key 2048
  # $ openssl req -x509 -new -nodes -key ca.key -days 10000 -out ca.crt -subj "/CN=mydomain.com"
  # After that you can base64 encode it and paste it here:
  # $ cat ca.key | base64 -w0
  # cacert:
  # cakey:

# Entries for the MongoDB config file
configmap:

# Javascript code to execute on each replica at initContainer time
# This is the recommended way to create indexes on replicasets.
# Below is an example that creates indexes in foreground on each replica in standalone mode.
# ref: https://docs.mongodb.com/manual/tutorial/build-indexes-on-replica-sets/
# initMongodStandalone: |+
#   db = db.getSiblingDB("mydb")
#   db.my_users.createIndex({email: 1})

# Readiness probe
readinessProbe:
  initialDelaySeconds: 5
  timeoutSeconds: 1
  failureThreshold: 3
  periodSeconds: 10
  successThreshold: 1

# Liveness probe
livenessProbe:
  initialDelaySeconds: 30
  timeoutSeconds: 5
  failureThreshold: 3
  periodSeconds: 10
  successThreshold: 1
```

Deployment works. But I cannot login to MongoDB replicaset. I'm using this command to login to Replica Set.
```
mongo mongodb://admin:passwordDB1231231123zzdd19@mongodb-replicaset:27017/?replicaSet=rs0
```

But it returns this:

```
root@nodetest# mongo mongodb://admin:passwordDB1231231123zzdd19@mongodb-replicaset:27017/?replicaSet=rs0
MongoDB shell version v3.6.8
connecting to: mongodb://mongodb-replicaset:27017/?replicaSet=rs0
2018-10-23T14:32:00.109+0000 I NETWORK  [thread1] Starting new replica set monitor for rs0/mongodb-replicaset:27017
2018-10-23T14:32:00.113+0000 I NETWORK  [thread1] Successfully connected to mongodb-replicaset:27017 (1 connections now open to mongodb-replicaset:27017 with a 5 second timeout)
2018-10-23T14:32:00.113+0000 I NETWORK  [thread1] changing hosts to rs0/mongodb-replicaset-0.mongodb-replicaset.default.svc.cluster.local:27017,mongodb-replicaset-1.mongodb-replicaset.default.svc.cluster.local:27017,mongodb-replicaset-2.mongodb-replicaset.default.svc.cluster.local:27017 from rs0/mongodb-replicaset:27017
2018-10-23T14:32:00.116+0000 I NETWORK  [ReplicaSetMonitor-TaskExecutor-0] Successfully connected to mongodb-replicaset-0.mongodb-replicaset.default.svc.cluster.local:27017 (1 connections now open to mongodb-replicaset-0.mongodb-replicaset.default.svc.cluster.local:27017 with a 5 second timeout)
2018-10-23T14:32:00.118+0000 I NETWORK  [thread1] Successfully connected to mongodb-replicaset-1.mongodb-replicaset.default.svc.cluster.local:27017 (1 connections now open to mongodb-replicaset-1.mongodb-replicaset.default.svc.cluster.local:27017 with a 5 second timeout)
2018-10-23T14:32:00.120+0000 I NETWORK  [ReplicaSetMonitor-TaskExecutor-0] Successfully connected to mongodb-replicaset-2.mongodb-replicaset.default.svc.cluster.local:27017 (1 connections now open to mongodb-replicaset-2.mongodb-replicaset.default.svc.cluster.local:27017 with a 5 second timeout)
MongoDB server version: 3.6.8
2018-10-23T14:32:00.123+0000 E QUERY    [thread1] Error: Authentication failed. :
DB.prototype._authOrThrow@src/mongo/shell/db.js:1608:20
@(auth):6:1
@(auth):1:2
exception: login failed
```

TLDR; How do I enable authentication in mongodb replica set and how do I connect to it using mongodb:// uri?

**Comment by steven-sheehy** [comment]
You need to authenticate to the admin database but connect to non-admin database.  Try this on the node that is primary (or secondaries, but writes won't work):

```shell
kubectl exec -it mongodb-replicaset-0 -- mongo mydb -u admin -p password--authenticationDatabase admin
```

If you're talking about from your application container, try using a URI like:

`mongodb://${db.username}:${db.password}@mongodb-replicaset:27017/mydb?authSource=admin&replicaSet=rs0`

**Comment by shinebayar-g** [comment]
Well I don't ssh into nodes directly. I'm just trying to login from Mongo shell from container living in same cluster of mongodb replica set.

`mongo --host rs0/mongodb-replicaset:27017 -u admin -p passwordDB1231231123zzdd19`
I am supposed to login to replica set using this command, but it says login failed and I don't know why.

**Comment by steven-sheehy** [comment]
> This command basically works, but it's not letting me to write any command. If I write anything here, it closes.

You need to login to the primary to write, as I previously stated. Replace mongodb-replicaset with the pod dns name that is the primary. It should be something like `mongodb-replicaset-0.mongodb-replicaset.default.svc.cluster.local`, replacing namespace, cluster domain, pod name and service name as appropriate.

**Comment by okgolove** [comment]
I have similar problem.

```
auth:
  adminPassword: test
  adminUser: test
  enabled: "true"
  key: biglongcoolkey
persistentVolume:
  enabled: true
resources:
  limits:
    memory: 256Mi
```

https://github.com/helm/charts/blob/master/stable/mongodb-replicaset/init/on-start.sh#L198
As I can see init script doesn't create new user:
`kl mongodb-replicaset-0 -c bootstrap | grep "on-start"`
```
2018/10/25 10:43:19 execing: /init/on-start.sh with stdin: mongodb-mongodb-replicaset-0.mongodb-mongodb-replicaset.default.svc.cluster.local
2018/10/25 10:44:05 [2018-10-25T10:43:19,098876414+00:00] [on-start.sh] Bootstrapping MongoDB replica set member: mongodb-mongodb-replicaset-0
[2018-10-25T10:43:19,109928265+00:00] [on-start.sh] Reading standard input...
[2018-10-25T10:43:19,113156641+00:00] [on-start.sh] Skipping init mongod standalone script
[2018-10-25T10:43:19,116151558+00:00] [on-start.sh] Peers: mongodb-mongodb-replicaset-0.mongodb-mongodb-replicaset.default.svc.cluster.local
[2018-10-25T10:43:19,118863617+00:00] [on-start.sh] Starting a MongoDB replica
[2018-10-25T10:43:19,123140193+00:00] [on-start.sh] Waiting for MongoDB to be ready...
[2018-10-25T10:43:19,456111258+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:20,724531677+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:21,941101355+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:23,113823529+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:24,337276965+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:25,599066624+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:26,880450656+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:28,074519806+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:29,490083889+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:30,766199601+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:32,111392250+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:33,278921484+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:34,613516802+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:36,006845183+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:38,105018961+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:39,409736287+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:40,760598440+00:00] [on-start.sh] Retrying db.adminCommand('ping').ok
[2018-10-25T10:43:41,965272093+00:00] [on-start.sh] Initialized.
[2018-10-25T10:43:41,973317759+00:00] [on-start.sh] Checking if mongodb-mongodb-replicaset-0.mongodb-mongodb-replicaset.default.svc.cluster.local is primary
[2018-10-25T10:43:42,485394979+00:00] [on-start.sh] Retrying db.isMaster().ismaster
[2018-10-25T10:43:44,141208413+00:00] [on-start.sh] Found primary: mongodb-mongodb-replicaset-0.mongodb-mongodb-replicaset.default.svc.cluster.local
[2018-10-25T10:43:44,148861626+00:00] [on-start.sh] This replica is already PRIMARY
[2018-10-25T10:43:44,158201179+00:00] [on-start.sh] MongoDB bootstrap complete
[2018-10-25T10:43:44,176220646+00:00] [on-start.sh] Shutting down MongoDB (force: true)...

```

**Comment by steven-sheehy** [comment]
> [2018-10-25T10:43:44,148861626+00:00] [on-start.sh] This replica is already PRIMARY

This indicates it has already been started previously. User creation only occurs the first time before any rs.init(). Delete your persistent volumes and restart.

**Comment by shinebayar-g** [comment]
My connection uri and method was right. But I had need to setup another few lines in values.yaml file which explained here https://github.com/helm/charts/issues/2976

**Comment by okgolove** [comment]
Yeah, I've already understood it. Thank you for your explanation!

**Comment by shinebayar-g** [comment]
Yea it's just shame that issue still open for year.
